linera-base/src/data_types/balance_scalar.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! A custom `ScalarType` implementation for `Balance`.
+//!
+//! GraphQL doesn't support `u128`, so we represent a balance as a list with two elements, the most
+//! and least significant 64 bits.
+//!
+//! This is mostly adapted from the implementation of the `async_graphql::scalar!` macro.
+
+use std::{borrow::Cow, iter, sync::Arc};
+
+use async_graphql::{
+    parser::types::Field,
+    registry::{MetaType, MetaTypeId, Registry},
+    ContextSelectionSet, InputType, InputValueError, InputValueResult, Name, OutputType,
+    Positioned, ScalarType, ServerResult, Value,
+};
+
+use super::Balance;
+
+const BALANCE_DESCRIPTION: &str = "The balance of a chain";
+const UPPER: &str = "upper";
+const LOWER: &str = "lower";
+
+impl ScalarType for Balance {
+    fn parse(value: Value) -> InputValueResult<Self> {
```

Comment:
My two (minor) issues with this PR are:
* because `scalar!` is based on `serde_json`, I believe one could just define `serde::{Serialize, Deserialize}` manually for `Balance` possibly using `if serializer.is_human_readable() { ... } else { ... }` to keep the binary encoding format as it is.
* If we don't do that, there should be unit tests for the hand-written (and nontrivial) parser and printer.

Reply:
Good idea, done.